packages/protocol/lib/backward/ast-code.ts

```diff
@@ -0,0 +1,323 @@
+// tslint:disable: max-classes-per-file
+import { makeZContract } from '@celo/protocol/lib/backward/internal'
+import {
+  BuildArtifacts,
+  Contract as ZContract
+} from '@openzeppelin/upgrades'
+import ContractAST from '@openzeppelin/upgrades/lib/utils/ContractAST'
+
+export const VISIBILITY_PUBLIC = 'public'
+export const VISIBILITY_EXTERNAL = 'external'
+const CONTRACT_TYPE_CONTRACT = 'contract'
+const STORAGE_DEFAULT = 'default'
+
+const OUT_VOID_PARAMETER_STRING = 'void'
+
+export class ASTCodeCompatibilityReport {
+  changes: Change[]
+  constructor(changes: Change[]) {
+    this.changes = changes
+  }
+  public push(...changes: Change[]) {
+    this.changes.push(...changes)
+  }
+  public include(other: ASTCodeCompatibilityReport) {
+    this.push(...other.changes)
+  }
+}
+
+export interface Change {
+  getContract(): string
+  accept<T>(visitor: ChangeVisitor<T>): T
+}
+
+export enum ChangeType { Patch, Minor, Major }
+
+export interface ChangeVisitor<T> {
+  onMethodMutability(change: MethodMutabilityChange): T
+  onMethodParameters(change: MethodParametersChange): T
+  onMethodReturn(change: MethodReturnChange): T
+  onMethodVisibility(change: MethodVisibilityChange): T
+  onMethodAdded(change: MethodAddedChange): T
+  onMethodRemoved(change: MethodRemovedChange): T
+  onContractType(change: ContractTypeChange): T
+  onNewContract(change: NewContractChange): T
+  onDeployedBytecode(change: DeployedBytecodeChange): T
+}
+
+abstract class ContractChange implements Change {
+  type: string
+  contract: string
+  constructor(contract: string) {
+    this.contract = contract
+  }
+  getContract() {
+    return this.contract
+  }
+
+  abstract accept<T>(visitor: ChangeVisitor<T>): T
+}
+
+export class NewContractChange extends ContractChange {
+  type = "NewContract"
```

Comment:
this looks strange.... type here is an instance variable. Not sure

Reply:
I needed a discriminator to describe what the change type is in the serialized json. TBH I don't like it either but I'm not sure how to accomplish this in a different way